### ISSUE TITLE
Bug: Parser does not report SyntaxError when encountering strict mode reserved words

### ISSUE THREAD

**Issue by eps1lon** [typescript-eslint/typescript-eslint#7720]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Playground Link

https://typescript-eslint.io/play#ts=5.2.2&fileType=.js&code=OQVwzgpgBGAuBOBLAxrYAoA9Jq8KXgDcIATKAdwHt4SwpEA7KAKQGUYEVYoBbSk6ACMQ3AIYAbcZXKkoggJ5QADqPgF0ySgzj0eS8RB4QGsOgF4owPAWIkMm7d0awI8AGajk0C1fyvb9lo6KsgA1qIA5t6W1v6kgY7KSISiLlA%2BsUTxGkHcSvCULqiyGX5ZdjmJSiCC4ijpMWUB6EA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYBfEWg&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
'use strict'
// reserved words in JS strict mode but allowed by parser
const implements = 'reserved'
const interface = 'reserved'
const package = 'reserved'
const private = 'reserved'
const protected = 'reserved'
const public = 'reserved'
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {},
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

SyntaxError for every identifier

### Actual Result

No SyntaxError

### Additional Info

There are [reserved words](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#reserved_words), and [reserved words in JS strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#future_reserved_words). Maybe the parser just doesn't know about reserved words in strict mode?

[TypeScript correctly raises a SyntaxError here](https://www.typescriptlang.org/play?#code/OQVwzgpgBGAuBOBLAxrYAoA9Jq8KXgDcIATKAdwHt4SwpEA7KAKQGUYEVYoBbSk6ACMQ3AIYAbcZXKkoggJ5QADqPgF0ySgzj0eS8RB4QGsOgF4owPAWIkMm7d0awI8AGajk0C1fyvb9lo6KsgA1qIA5t6W1v6kgY7KSISiLlA+sUTxGkHcSvCULqiyGX5ZdjmJSiCC4ijpMWUB6EA), so does Node.js and [Babel.js](https://astexplorer.net/#/gist/f6bc57d68691ab59dd72ad7d5000b0f6/e9a04637c24db3e9e65861e08ca3eed3ae826199)

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `6.7.4` |
| `@typescript-eslint/parser`            | `6.7.4` |
| `@typescript-eslint/rule-tester`       | `6.7.4` |
| `@typescript-eslint/scope-manager`     | `6.7.4` |
| `@typescript-eslint/typescript-estree` | `6.7.4` |
| `@typescript-eslint/type-utils`        | `6.7.4` |
| `@typescript-eslint/utils`             | `6.7.4` |
| `TypeScript`                           | `5.2.2` |
| `ESLint`                               | `8.48.0` |
| `node`                                 | `18.7.0` |

**Comment by bradzacher** [comment]
> [TypeScript correctly raises a SyntaxError here](https://www.typescriptlang.org/play?#code/OQVwzgpgBGAuBOBLAxrYAoA9Jq8KXgDcIATKAdwHt4SwpEA7KAKQGUYEVYoBbSk6ACMQ3AIYAbcZXKkoggJ5QADqPgF0ySgzj0eS8RB4QGsOgF4owPAWIkMm7d0awI8AGajk0C1fyvb9lo6KsgA1qIA5t6W1v6kgY7KSISiLlA+sUTxGkHcSvCULqiyGX5ZdjmJSiCC4ijpMWUB6EA)

You're half correct - TS raises a ***semantic*** error here - not a *syntax* error.
What's the difference?

A syntax error occurs within TS's parser if and only if the parser encounters syntax with an unrecoverable error. For example TS emits a syntax error for code like `class 'Foo' {}`.

A semantic error occurs in all other cases - including non-syntactic error cases (eg type errors).

We do not expose semantic errors from our parser because of this last fact - in order to calculate them we also need to ask TS to calculate all type errors for the entire `tsconfig.json`. There is no in-between here. This is a prohibitively expensive operation - in some codebases it can take 50ms per file even if there are no errors! That's a lot of time to waste returning zero results!

Instead we expose all syntax errors TS reports and we expose our own curated set of errors that we are slowly building up. With our custom errors we are specifically looking to add errors in cases where the AST won't be correct or will be non-sensical (for example `var;` will error with our own error and TS will emit a semantic error).

I don't think we care about enforcing cases like this - TS will semantic error - meaning your code will fail its typecheck, and consumers like ESLint or Prettier don't care about "bad" identifiers (as long as they match the allowed identifier pattern the exact name isn't important). So there's not really good reason for us to specifically error here.

**Comment by bradzacher** [comment]
Our philosophy is that our parser is built on top of TS and ESLint is intended to complement TS.

So from our POV we don't want to validate the AST is 100% correct, no, because we trust that TS will check that and enforce the user is writing correct code.

It wasn't until our last major that we added the first of our custom errors! Before that we only errored specifically on TS syntax errors.

The errors we want are errors which enforce the AST object matches the types and satisfies certain invariants that are encoded in the ESTree spec.

So in this specific case - there is one identifier and it matches the allowed identifier pattern - so it satisfies the invariants we care about.

Deeper validation can be provided by TS at typecheck time outside of the lint run.

**Comment by eps1lon** [comment]
Raised a TS issue to confirm. I expect my linter to catch syntactic errors so this is a clear bug to me.

**Comment by bradzacher** [comment]
> I expect my linter to catch syntactic errors so this is a clear bug to me.

To be clear - never in the history of our project (>5y) have we ever errored on this code.

This isn't a recent regression or change - it is and always has been the way our tooling works.

**Comment by JoshuaKGoldberg** [comment]
> I expect my linter to catch syntactic errors so this is a clear bug to me.

Out of curiosity, are there other linters -in or out of the TypeScript ecosystem, or even JavaScript as a whole- that do or don't have opinions on this?

ESLint's built-in parser does produce a parsing error on each of them: https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgaW1wbGVtZW50cyA9ICdyZXNlcnZlZCdcbmNvbnN0IGludGVyZmFjZSA9ICdyZXNlcnZlZCdcbmNvbnN0IHBhY2thZ2UgPSAncmVzZXJ2ZWQnXG5jb25zdCBwcml2YXRlID0gJ3Jlc2VydmVkJ1xuY29uc3QgcHJvdGVjdGVkID0gJ3Jlc2VydmVkJ1xuY29uc3QgcHVibGljID0gJ3Jlc2VydmVkJyIsIm9wdGlvbnMiOnsiZW52Ijp7ImVzNiI6dHJ1ZX0sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19

```plaintext
const implements = 'reserved'
//    ~
// Parsing error: The keyword 'implements' is reserved
```

**Comment by eps1lon** [comment]
> This isn't a recent regression or change - it is and always has been the way our tooling works.

I didn't want to give that impression. 

But from a user perspective it is a change in behavior if we you switch from ESLint + default parser to ESLint + TSESlint parser as Josh pointed out. That's where my intuition for "linters catch syntax errors" comes from. Maybe this was only incidental and the idea was always to be able to recover from as many as possible. 

If you don't want TSESlint to be a syntax checker, then so be it. But I would like to have some in-editor/CI tooling that catches syntax errors as early as possible.

**Comment by bradzacher** [comment]
> But from a user perspective it is a change in behavior if we you switch from ESLint + default parser to ESLint + TSESlint parser as Josh pointed out.

We are a parser for TS code - so the only way you'd make this change is if you're using our parser for JS code. It'll fully work for sure - but it's a sledgehammer to crack a nut.

---

Other parsers (like espree, eslint's default parser) are built to be very unambiguous and they generally follow the JS spec to the letter so it will disallow code that will syntactically crash at runtime.

For JS code this makes sense because, well, there isn't any other tooling that exists to check your code! If you don't run a linter you're probably only relying on tests to do runtime validation of your syntax - which is sketchy!

But for TS code as mentioned the linter is a complement to TS - so with or without the linter you always are running TS to check your code. Hence it's never been a priority or even a want for us to error on all "bad" code - because there's guaranteed to be another tool that you're running that does that check.
